forem/forem
# Refactor the home feed + add user tag filtering on in-feed billboards

## What type of PR is this? (check all applicable)


## Description

The billboard tag list should accept the user's followed tags for targeting. Hence, on the home feed we should be able target by the users followed tags. 

- If a Display Ad has no tags, it is not filtered on any tags, and we can display any ad on the home feed that meet the other filters. 
- If a Display Ad has tags, then it will only show on the home feed if the user follows any of those tags. 
- For example: If a display ad has tags #java and #career, and the user following tags include #java then that ad will be a contender for that placement area, along with all the the display ads that have no tags or are tagged with #java OR #career.
- However if a display ad has tags #java and #career, and the user following tags does not include any of those then that ad will not be a contender for that placement area.

## Related Tickets & Documents

- Related Issue #https://github.com/forem/forem/issues/19313
- Closes #

## QA Instructions, Screenshots, Recordings
Since I've implemented a refactor to the home feed, we'd want to re-test that the feed renders appropriately. 

### Test the Feed again after refactor: 

❗️ Its important to do thorough testing on this PR as it changes the implementation of the signed in home feed 

A loom Demo to demo and explain how to test:
https://www.loom.com/share/35c560f9a6ae40038fe0993712fb1bef

Note that you want to make sure that scrolling, changing between timeframes and everything else on the feed works as expected. If you test on Uffizi, you will not be able to probably get podcasts to show up. 


We implement the following organization for the feed:
```
1. Place the pinned post first
2. Place the image post next
5. Place the podcast episodes of a podcast that you follow which has new episode out today
6. Place the rest of the stories for the feed
7. Insert the billboards in that array accordingly
- feed_first: Before all home page posts
- feed_second: Between 2nd and 3rd posts in the feed
- feed_third: Between 7th and 8th posts in the feed
```

In order to test this you want to make sure that you have all the items setup in the list above: 

1. Pin a post 
Find a pinned post:
```
pinned_article = PinnedArticle.get
```
if there isn't one then run: 
```
PinnedArticle.set(Article.first)
```

2. Set a post with an image
```
articles = Articles::Feeds::Basic.new(user: User.find(11), page: @page).more_comments_minimal_weight_randomized.to_a
first_article = articles.first
first_article.update!(main_image: "https://pigment.github.io/fake-logos/logos/medium/color/2.png")
```

3. Setup your podcasts: We want to ensure that we have a followed podcasts that has episodes published today so that they can show up on the feed.

```
podcast = Podcast.first

user = <the user you're logged in with>
user.follow podcast

Podcasts::GetEpisodesWorker.perform_async("podcast_id" => podcast.id, "limit" => 10)
podcast_episode = podcast.podcast_episodes
podcast_episodes.limit(4).update(published_at: DateTime.now)
```

No need to do anything for (4).

5. Add some billboards:

```
DisplayAd.create!(
  name: "#{Faker::Lorem.word} feed_first",
  body_markdown: Faker::Lorem.sentence,
  published: true,
  approved: true,
  placement_area: "feed_first"
)
```

```
DisplayAd.create!(
  name: "#{Faker::Lorem.word} feed_second",
  body_markdown: Faker::Lorem.sentence,
  published: true,
  approved: true,
  placement_area: "feed_second"
)
```

```
DisplayAd.create!(
  name: "#{Faker::Lorem.word} feed_third",
  body_markdown: Faker::Lorem.sentence,
  published: true,
  approved: true,
  placement_area: "feed_third"
)
```

You can update the configurations by removing and adding some items and they should still maintain the order.

### Now, test the filtering by tags:

1. Go to http://localhost:3000/admin/customization/display_ads and create a new Ad.  When you click on the any of the home feed positions, you should see the Targeting tag field reveal itself. 

<img width="1232" alt="Screenshot 2023-06-12 at 11 37 54" src="https://github.com/forem/forem/assets/2786819/1de01b91-b698-41ee-8882-40bb56a6f050">

2. Add some tags to the Targeting field. Before you do, follow similar tags and take note of the users followed tags for the user you're logged in with `<current_user>.cached_followed_tag_names`


3. Test according to these rules: 
- If a Display Ad has no tags, it is not filtered on any tags, and we can display any ad on the home feed that meet other targeting requirements. 
- If a Display Ad has tags, then it will only show on the home feed if the user follows any of those tags. 
For example: If a display ad has tags #java and #career, and the user following tags include #java then that ad will be a contender for that placement area, along with all the the display ads that have no tags or are tagged with #java OR #career.
However if a display ad has tags #java and #career, and the user following tags does not include any of those then that ad will not be a contender for that placement area.

https://www.loom.com/share/a6cfe8f538194af5a9908137b8f07fed?sid=7a440095-9bac-447b-a7a2-af14304e4374

### UI accessibility checklist
_If your PR includes UI changes, please utilize this checklist:_

None, no real UI changes. 

## Added/updated tests?

      have not been included_

## [optional] Are there any post deployment tasks we need to perform?
We need to keep an eye on this PR after deployment as we implement a refactor to the home feed.

## [optional] What gif best describes this PR or how it makes you feel?

![alt_text](https://media.giphy.com/media/xTiN0CNHgoRf1Ha7CM/giphy.gif)
